### PR TITLE
Feature/rviz media export

### LINKED ISSUE #4: [rviz_rendering] implement and document replacement for media_export
In rviz in ROS 1, a package called `media_export` was used in combination with the `ROS_PACKAGE_PATH` via the `rospack` C++ API to allow other packages to include resources for Ogre to be used with their rviz plugins.

That code has been commented out in the ros2 port, but we need to find a suitable alternative for ROS 2. Again, this means using `ament_index` in a similar way as described in https://github.com/ros2/rviz/issues/3, but instead of using it to find one package you would use it to find packages which provide media for Ogre by looking at a different ament index resource, something like `rviz_ogre_media_exports` or similar. For details on ament index and how it could be used to do this, see resources linked to in https://github.com/ros2/rviz/issues/3.

### PR DESCRIPTION
**Pull request by Martin-Idel-SI**
This PR resolves #4 in the following way:

- Resources can be added as rviz_ogre_media_exports to the ament_index
- For convenience, a simple cmake macro taking the directories to be added as resources is provided
- RViz queries the ament_index at runtime adding all resources it finds
- The user is informed, if some of the claimed resource folders are missing

### CODE REVIEW

**Comment by wjwwood** [comment]
Sorry, I didn't think to do this one first, but the style pr I just merged means this needs to be rebased now.

File: rviz_rendering/CMakeLists.txt [hunk 1]
```diff
@@ -47,7 +48,7 @@ endif()
 # These need to be added in the add_library() call so AUTOMOC detects them.
 set(headers_with_q_object
   include/rviz_rendering/render_window.hpp
-)
+  src/rviz_rendering/string_helper.hpp)
```
Comment (wjwwood) [inline rviz_rendering/CMakeLists.txt]: Leave the `)` on the new line, the whole point is to simplify the diff :)

File: rviz_rendering/src/cmake/register_rviz_ogre_media_exports.cmake [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    set(RESOURCE_FILE_CONTENT "")
+    foreach(DIR ${ARGUMENTS_DIRECTORIES})
+        if(NOT IS_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/${DIR}")
+            message(FATAL_ERROR "Directory ${DIR} does not exist.
+                Paths must be specified relative to CMAKE_CURRENT_SOURCE_DIR")
+        endif()
+        set(RESOURCE_FILE_CONTENT "${RESOURCE_FILE_CONTENT}${PROJECT_NAME}/${DIR}\n")
+    endforeach()
+
+    ament_index_register_resource(rviz_ogre_media_exports CONTENT ${RESOURCE_FILE_CONTENT})
```
Comment (wjwwood) [inline rviz_rendering/src/cmake/register_rviz_ogre_media_exports.cmake]: Because this is written within this function, you should make it clear that this function can only be called once in the doc string.

A different way to do this would be to have each call to this function build up the content and then have a hook that runs in `ament_package()` which actually writes the resource.

See the other `ament_*` functions for examples:

- adds to variable in each call: https://github.com/ament/ament_cmake/blob/master/ament_cmake_core/cmake/environment_hooks/ament_environment_hooks.cmake#L78
- triggered via a hook: https://github.com/ament/ament_cmake/blob/36d5c721f5d9b6242126175e9002990017f849b0/ament_cmake_core/ament_cmake_environment_hooks-extras.cmake#L25-L26
- and consumed in macro called from hook file: https://github.com/ament/ament_cmake/blob/master/ament_cmake_core/cmake/environment_hooks/ament_generate_package_environment.cmake#L54

File: rviz_rendering/src/rviz_rendering/render_system.hpp [hunk 3]
```diff
@@ -52,6 +49,11 @@
 #include <OgreOverlaySystem.h>
 #include <RenderSystems/GL/OgreGLPlugin.h>
 
+#include <QDir>
+
+#include <cstdint>
+#include <string>
```
Comment (wjwwood) [inline rviz_rendering/src/rviz_rendering/render_system.hpp]: The system include should happen first, I know that makes cpplint unhappy, but that's the right order to do them in. I just disable the check on the Qt includes (on which it complains). This is the wrong order in my opinion, based on the list here:

https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

**Comment by Martin-Idel-SI** [comment]
@wjwwood No problem. I did the rebase.

**Comment by wjwwood** [comment]
Sorry forgot to post my comments from yesterday until just now :/

File: rviz_rendering/src/cmake/register_rviz_ogre_media_exports.cmake (hunk 2 above)
Comment (Martin-Idel-SI) [reply rviz_rendering/src/cmake/register_rviz_ogre_media_exports.cmake]: If your other proposition is the "usual" way of doing things for ament, then I'd prefer to also go this way to avoid future confusion. I'll have a look at it.

**Comment by Martin-Idel-SI** [comment]
I implemented a hook similar to the files you pointed out. 

Now, if I have a simple ros2 package, I can provide folders to rviz_ogre_media at will and they will be collected and installed at the end. For instance a call like this: 

`find_package(rviz_rendering REQUIRED)`
`register_rviz_ogre_media_exports(DIRECTORIES "ogre_media")`
`register_rviz_ogre_media_exports(DIRECTORIES "second_media")`
`register_rviz_ogre_media_exports(DIRECTORIES "second_media")`

would install the folders `ogre_media` and `second_media` (the latter twice, but the second time, it's clearly up-to-date) with `ament_package()`. Please review again, thank you!

File: rviz_rendering/src/rviz_rendering/render_system.hpp [hunk 4]
```diff
@@ -52,6 +49,11 @@
 #include <OgreOverlaySystem.h>
 #include <RenderSystems/GL/OgreGLPlugin.h>
 
+#include <cstdint>
+#include <string>
+
+#include <QDir> // NOLINT
```
Comment (wjwwood) [inline rviz_rendering/src/rviz_rendering/render_system.hpp]: These should be included before the Ogre headers and outside of the warnings pragma. I would just move them back...

File: rviz_rendering/src/rviz_rendering/string_helper.hpp [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+#include <vector>
+
+namespace rviz_rendering
+{
+
+class string_helper
+{
+public:
+  static std::vector<std::string>
+  splitStringIntoTrimmedItems(const std::string &, const char);
```
Comment (wjwwood) [inline rviz_rendering/src/rviz_rendering/string_helper.hpp]: Why does this need to be a static method of a class? Would it not be just as good as a free function in a namespace?

It seems odd to be able to instantiate a `string_helper` class which can do nothing special.

**Review by wjwwood — COMMENTED** [review]
It would be great to have a test for this feature, though I know it would require another package. I could imagine having a package embedded within this one which is built as part of the tests, but it might be just as well to have an example package which demonstrates how this works as well as tests it.

**Comment by Martin-Idel-SI** [comment]
Yes, I was also thinking about the testing issue. 

I had a small example set up locally, so I'll clean this up and add automated tests of the functionality. Since at some point, we need a developer documentation, we will probably want to have a rather minimal example plugin anyway, so maybe it's a good idea to start setting this up right now. We could then evolve the example into a true rviz example plugin in due course.

**Comment by Martin-Idel-SI** [comment]
I updated the PR with requested changes and new tests testing the macro (all tests are passing on Linux at least). I decided on an extra package, but of course, we can change this (especially the name).

File: rviz_rendering/CMakeLists.txt [hunk 6]
```diff
@@ -126,6 +129,12 @@ if(TARGET Qt5::windeployqt)
     )
 endif()
 
+set(rviz_rendering_DIR "${CMAKE_CURRENT_SOURCE_DIR}/src/cmake")
+include(${CMAKE_CURRENT_SOURCE_DIR}/src/cmake/register_rviz_ogre_media_exports_hook-extras.cmake)
+
+install(DIRECTORY src/cmake
+        DESTINATION share/${PROJECT_NAME})
```
Comment (wjwwood) [inline rviz_rendering/CMakeLists.txt]: Please don't do aligned indentations (c++ style check will catch this, but none for CMake). Instead do:

```cmake
install(DIRECTORY src/cmake
  DESTINATION share/${PROJECT_NAME})
```

Or:

```cmake
install(
  DIRECTORY src/cmake
  DESTINATION share/${PROJECT_NAME})
```

Please check the rest of the files, I'm seeing it a lot.

File: rviz_rendering/CMakeLists.txt [hunk 7]
```diff
@@ -142,6 +151,16 @@ if(BUILD_TESTING)
...
+        target_link_libraries(test_rviz_rendering ${PROJECT_NAME})
+    endif()
 endif()
 
-ament_package()
+list(APPEND ${PROJECT_NAME}_CONFIG_EXTRAS
+        "${CMAKE_CURRENT_SOURCE_DIR}/src/cmake/register_rviz_ogre_media_exports_hook-extras.cmake"
+        "${CMAKE_CURRENT_SOURCE_DIR}/src/cmake/register_rviz_ogre_media_exports.cmake")
+
+ament_package()
```
Comment (wjwwood) [inline rviz_rendering/CMakeLists.txt]: Please make sure all the files have a newline at the end of the file (again the C++ style checker would catch this). You can see it on github like this:

screen shot 2017-10-25 at 8 24 46 pm

File: rviz_rendering/src/cmake/register_rviz_ogre_media_exports_hook-extras.cmake [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+option(AMENT_RVIZ_OGRE_MEDIA
+        "Add ogre_media resources to RViz" ON)
```
Comment (wjwwood) [inline rviz_rendering/src/cmake/register_rviz_ogre_media_exports_hook-extras.cmake]: This is neither aligned indention nor the standard two-space indention. I see this a lot in the file too.

File: rviz_rendering/src/rviz_rendering/render_system.cpp [hunk 9]
```diff
@@ -39,21 +40,25 @@
 #include <GL/glx.h>
 #endif
 
+
```
Comment (wjwwood) [inline rviz_rendering/src/rviz_rendering/render_system.cpp]: Please avoid unnecessary whitespace changes and minimize vertical whitespace: https://google.github.io/styleguide/cppguide.html#Vertical_Whitespace

File: rviz_rendering/src/rviz_rendering/render_system.cpp [hunk 10]
```diff
@@ -330,32 +335,34 @@ RenderSystem::setupResources()
...
+    RVIZ_OGRE_MEDIA_RESOURCE_NAME);
+  for (auto resource : resource_locations) {
+    std::string content;
+    std::string prefix_path;
+    if (ament_index_cpp::get_resource(RVIZ_OGRE_MEDIA_RESOURCE_NAME, resource.first, content,
+      &prefix_path))
+    {
+      std::vector<std::string> filenames =
+        string_helper::splitStringIntoTrimmedItems(
+        content, '\n');
```
Comment (wjwwood) [inline rviz_rendering/src/rviz_rendering/render_system.cpp]: nitpick: I would restyle this to be:

```c++
      std::vector<std::string> filenames =
        string_helper::splitStringIntoTrimmedItems(content, '\n');
```

Or:

```c++
      std::vector<std::string> filenames = string_helper::splitStringIntoTrimmedItems(
        content, '\n');
```

But not both. You don't need to fix this particular case, but just for future reference.

File: rviz_rendering/src/rviz_rendering/render_system.hpp [hunk 11]
```diff
@@ -56,6 +56,8 @@
 # pragma GCC diagnostic pop
 #endif
 
+#include <QDir> // NOLINT
```
Comment (wjwwood) [inline rviz_rendering/src/rviz_rendering/render_system.hpp]: Two spaces between comment and code please:

```c++
#include <QDir>  // NOLINT
```

File: rviz_rendering_tests/package.xml [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+    <name>rviz_rendering_tests</name>
+    <version>0.0.1</version>
+    <description>
+        Example plugin for RViz - documents and tests RViz plugin development
+    </description>
+    <maintainer email="william@osrfoundation.org">William Woodall</maintainer>
+    <license>BSD</license>
+
+    <!-- TODO(Martin-Idel-SI): website does not exist -->
+    <url type="website">http://ros.org/wiki/rviz_rendering_tests</url>
```
Comment (wjwwood) [inline rviz_rendering_tests/package.xml]: Just leave this out for now rather than TODO it.

**Review by wjwwood — CHANGES_REQUESTED** [review]
I had a few style comments, which btw I would have taken care of for you, but I don't seem to have push permission on this pr for some reason. (note I don't need it on the repository, just the branch and the pr normally does this for you unless you opt out when creating it)

I'm ok with the new package right now. Later we might roll it into a rviz plugin example package where we can just as easily do the same tests.

**Comment by Martin-Idel-SI** [comment]
Hopefully fixed all code-style issues. It's good to bring them up - I'm relying on my IDE + linting to handle the code style and apparently I still had some problems with the configuration. 

By the way, in order to simplify reviewing, I'm not squashing. If you want me to squash beforehand, I can do that.

**Comment by Martin-Idel-SI** [comment]
I also had to remove the anonymous namespace, since it makes cppcheck >1.78 fail for some unknown reason.

**Comment by Martin-Idel-SI** [comment]
I rebased everything onto the current master and hopefully all requested changes are implemented.

File: rviz_rendering/CMakeLists.txt [hunk 13]
```diff
@@ -107,41 +110,57 @@ target_link_libraries(rendering_example rviz_rendering Qt5::Widgets)
...
 if(TARGET Qt5::windeployqt)
-    # execute windeployqt in a tmp directory after build
-    add_custom_command(TARGET rendering_example
-        POST_BUILD
-        COMMAND ${CMAKE_COMMAND} -E remove_directory "${CMAKE_CURRENT_BINARY_DIR}/windeployqt"
-        COMMAND set PATH=%PATH%$<SEMICOLON>${qt5_install_prefix}/bin
-        COMMAND
-          Qt5::windeployqt
-          --dir "${CMAKE_CURRENT_BINARY_DIR}/windeployqt"
-          "$<TARGET_FILE_DIR:rendering_example>/$<TARGET_FILE_NAME:rendering_example>"
```
Comment (wjwwood) [inline rviz_rendering/CMakeLists.txt:118]: Re-indenting to two space is correct, but the extra indentation for the command's arguments should have stayed. You don't have to fix this, but for the future reference.

File: rviz_rendering/src/rviz_rendering/string_helper.cpp [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+  std::stringstream stringstream(string_to_split);
+  std::string item;
+  std::vector<std::string> filenames;
+  while (std::getline(stringstream, item, delimiter)) {
+    auto whitespace_front = std::find_if_not(item.begin(), item.end(), [](char character) {
+          return std::isspace<char>(character, std::locale(""));
+        });
+    auto whitespace_back = std::find_if_not(item.rbegin(), item.rend(), [](char character) {
+          return std::isspace<char>(character, std::locale(""));
+        });
```
Comment (wjwwood) [inline rviz_rendering/src/rviz_rendering/string_helper.cpp:51]: The indentation doesn't look right here. I'll check it tomorrow morning, but make sure you have the latest ros2 sources I think this was fixed recently in our fork of uncrustify.

**Review by wjwwood — APPROVED** [review]

**Comment by wjwwood** [comment]
I'll redo the uncrustify later.

File: rviz_rendering/src/rviz_rendering/string_helper.cpp (hunk 14 above)
Comment (Martin-Idel-SI) [reply rviz_rendering/src/rviz_rendering/string_helper.cpp:51]: I just updated my ros2 installation but uncrustify doesn't seem to complain. I agree that it's not the best, though.

Comment (wjwwood) [reply rviz_rendering/src/rviz_rendering/string_helper.cpp:51]: Hmm, that sucks. Maybe not all the cases involving lambda's is fixed. Oh well, better to have it consistent with the tool, but it's always frustrating when the linter makes us do something we don't want 💢.
